go/chat/signencrypt/codec.go

```diff
@@ -0,0 +1,530 @@
+// This is a construction for encrypting and signing a message, using a
+// symmetric encryption key and a signing keypair, in a way that supports safe
+// streaming decryption. We need this for chat attachments because we've chosen
+// to use signing keys for authenticity in chat, and we don't want one
+// participant to be able to modify another's attachment, even with an evil
+// server's help. It's *almost* enough that we record the hash of the
+// attachment along with the symmetric key used to encrypt it, but that by
+// itself doesn't allow safe streaming decryption. Instead, we use this
+// construction to sign each chunk of the attachment as we encrypt it. (Note
+// that it's still possible for a sender with the server's help to modify their
+// *own* attachments after the fact, if clients aren't checking the hash. This
```

Comment:
No, the idea is that you change the attachment file itself _without_ changing the message that refers to it, so prev pointers don't help you. The message asserts `(encryption key, signing key, hash)`. Of course if you check the hash then there's no flexibility on the attachment bytes at all. But if you _don't_ (because you're streaming) than you'll accept _any_ file that was signed by the right keys. There could be more than one, if the server allows it.